### DISCUSSION TITLE
Value replacement expression; generalized form of ??

### DISCUSSION THREAD

**Discussion by dylanrafael05** [dotnet/csharplang#4594] [Language Ideas] [answered]
### Preamble
Very simple idea, and I have no idea for a good syntax but consider the following case:
```cs
// someValue is a parameter in a method and is of type int
someValue = (someValue == -1 ? 100 : someValue);
```
Although this can be condensed using:
```cs
// someValue is a parameter in a method and is now of type (int?)
someValue ??= 100;
```
The type of the variable has now been changed, meaning direct access to the value of `someValue` has now been removed.
### Proposal
In order to mitigate this, a "value substitution operator" could be implemented:
```cs
// someValue is a parameter in a method and is of type int

someValue = (someValue !! -1 ? 100);
// Value replacement syntax; parens are not required.
// (a !! b ? c) evaluates to (a == b ? c : a)
```
When replacing the default value of a type, `b` may be ommitted:
```cs
// someValue = (someValue !! default ? 100);
someValue = (some !! ? 100);
```
### Example code
```cs
public float SafeFloatDivide(float num, float den) 
{
    den = (den !! ? 0.01f);
    return num / den;
}

public string EmptyCheck(string str) 
{
    return str !! "" ? "(empty)";
}
```
Could this be useful / common enough to be considered?
I believe this pattern is especially common when a default parameter is of a value type and the desired default value is not constant.

**Comment by HaloFour** [comment, marked as answer]
IMO the ternary conditional operator is sufficiently terse in these cases.

**Reply by dylanrafael05** [reply]
Couldn't it be said that `objVar = (objVar == null ? expr : objVar);` is also sufficiently terse?
Furthermore, the examples that I gave were not optimal in demonstrating usefulness, but instead demonstrate the possible uses of such an operator.

**Reply by HaloFour** [reply]
Possibly, members of the LDG were pretty skeptical of adding `??=` until the community was able to demonstrate enough use cases to justify the form of the existing operator.  The LDG has also declined to consider compound assignment versions of other existing operators, like `&&` and `||`.  It's about the value-add proposition and reward on investment, and these operators feel pretty niche to me which makes me believe that they wouldn't be used often enough to justify their addition.

**Reply by dylanrafael05** [reply]
In that case, just the syntax for `!!` should be considered. `!!=` is far more niche and easily replaceable with a `!!` expression, and could be left out or implemented later like `??=`

Edit: _Removed !!=, !?, and !?= from proposal in order to gauge reasonability_

**Reply by CyrusNajmabadi** [reply]
I find this much more difficult to read or understand. Can you provide some real world cases to help gauge this?

**Reply by dylanrafael05** [reply]
> I find this much more difficult to read or understand. Can you provide some real world cases to help gauge this?

_Typing on mobile, excuse spelling errors_
I came up with this idea while working with a method that found an item within an indexable type.
```cs
public object ItemAt
(
   CustomCollection collect,
   CustomPath path,
   int? end = null
) {
   int endReal = end ?? collect.Length;
   //...
}
```
Boxing the default parameter "end" felt clunky, since now there's a new `int?` in memory that needs to be cleaned up. Although it is possible to create an overloaded method without the end parameter that does this calculation, the amount of extra code needed (a new method, a method body, and other boilerplate) seems unnecessary, and such a solution may even be impossible, especially if the desired default value depends on code executed within the function. Various other proposals (none of which I can find at the moment) suggest that a good solution to this would be the allowance of non-constant default parameters, and while that would be a better solution, the creation of a terse operator to handle value substitution allows for basically the same thing:
```cs
public object ItemAt
(
   CustomCollection collect,
   CustomPath path,
   int end = -1
) {
   end = (end !! -1 ? collect.Length);
   //...
}
``` 
... provided that there is some value of the argument that can be interpreted as invalid. This operator can also be used to streamline simple, ternary-like operations, since it abstracts away the need to repeat the variable name twice. For example: `str !! "foo" : "bar"` displays only the necessary information to deduce that the value of str is being replaced with "bar" if it is "foo", while `str == "foo" ? bar : str` reads more as an if statement using str in the conditional. While this is only a small difference, I believe the frequency at which such a pattern appears warrants a more concise way to denote value replacement, even if the difference is small.

**Reply by HaloFour** [reply]
> Boxing the default parameter "end" felt clunky, since now there's a new `int?` in memory that needs to be cleaned up.

Nullable value types are not reference types, there is no "boxing" or "cleaning up" involved.

I find the following much more readable:

```cs
public object ItemAt
(
   CustomCollection collect,
   CustomPath path,
   int end = -1
) {
   if (end == -1) end = collect.Length;
   // ...
}
```

Although I'd probably prefer to go with an overload here.

**Reply by theunrepentantgeek** [reply]
> Although it is possible to create an overloaded method without the end parameter that does this calculation, the amount of extra code needed (a new method, a method body, and other boilerplate) seems unnecessary,

The overhead doesn't seem very high for me - and it gives you considerable flexibility over what exactly happens:

``` csharp
public object ItemAt(
   CustomCollection collect,
   CustomPath path,
   int end) 
{
   //...
}

public object ItemAt(
   CustomCollection collect,
   CustomPath path) => ItemAt(collect, path, collect.Length)
```

**Reply by dylanrafael05** [reply]
Both valid answers, marking this post as answered and will not update in the future

**Reply by Tinister** [reply]
A generalized form of ?? might work in situations where you don't want to create a new variable.  e.g. `str.Substring(0, str.IndexOf(':') ?? str.Length)`
